src/Npgsql/Internal/Converters/RangeConverter.cs

```diff
@@ -0,0 +1,189 @@
+using System;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using NpgsqlTypes;
+
+namespace Npgsql.Internal.Converters;
+
+sealed class RangeConverter<T> : PgStreamingConverter<NpgsqlRange<T>>
+{
+    readonly PgConverter<T> _subtypeConverter;
+    readonly Size _subtypeBufferReadRequirements;
+
+    public RangeConverter(PgConverter<T> subtypeConverter)
+    {
+        if (!subtypeConverter.CanConvert(DataFormat.Binary, out var bufferingRequirement))
+            throw new NotSupportedException("Range subtype converter has to support the binary format to be compatible.");
+        (_subtypeBufferReadRequirements, _) = bufferingRequirement.ToBufferRequirements(DataFormat.Binary, subtypeConverter);
+
+        _subtypeConverter = subtypeConverter;
+    }
+
+    public override NpgsqlRange<T> Read(PgReader reader)
+        => Read(async: false, reader, CancellationToken.None).GetAwaiter().GetResult();
+
+    public override ValueTask<NpgsqlRange<T>> ReadAsync(PgReader reader, CancellationToken cancellationToken = default)
+        => Read(async: true, reader, cancellationToken);
+
+    async ValueTask<NpgsqlRange<T>> Read(bool async, PgReader reader, CancellationToken cancellationToken)
+    {
+        if (reader.ShouldBuffer(sizeof(byte)))
+            await reader.BufferData(async, sizeof(byte), cancellationToken).ConfigureAwait(false);
+
+        var flags = (RangeFlags)reader.ReadByte();
+        if ((flags & RangeFlags.Empty) != 0)
+            return NpgsqlRange<T>.Empty;
+
+        var lowerBound = default(T);
+        var upperBound = default(T);
+
+        if ((flags & RangeFlags.LowerBoundInfinite) == 0)
+        {
+            if (reader.ShouldBuffer(sizeof(int)))
+                await reader.BufferData(async, sizeof(int), cancellationToken).ConfigureAwait(false);
+            var length = reader.ReadInt32();
+
+            // Note that we leave the CLR default for nulls
+            if (length != -1)
+            {
+                // Set size before calling ShouldBuffer (it needs to be able to resolve an upper bound requirement)
+                reader.Current.Size = length;
```

Comment:
API-wise, it's a bit odd to have to set the size before calling ShouldBuffer - the relationship isn't clear. Maybe we should have a single ShouldBufferNewValue (or whatever) that also gets the length and sets Current.Size internally?

Reply:
Yeah 100%, I don't like the public mutability of size.

Ideally anything we come up with is a form of scope that can be started.

Reply:
Yep, exactly

Reply:
Will tackle this in one go, maybe during the multirange work